rules/internal/utils/http.go

```diff
@@ -26,21 +26,69 @@ import (
 // and then flattens the values in `additional_bindings`.
 // This allows rule authors to simply range over all of the HTTP rules,
 // since the common case is to want to apply the checks to all of them.
-func GetHTTPRules(m *desc.MethodDescriptor) []*apb.HttpRule {
-	rules := []*apb.HttpRule{}
+func GetHTTPRules(m *desc.MethodDescriptor) []*HTTPRule {
+	rules := []*HTTPRule{}
 
 	// Get the method options.
 	opts := m.GetMethodOptions()
 
 	// Get the "primary" rule (the direct google.api.http annotation).
 	if x, err := proto.GetExtension(opts, apb.E_Http); err == nil {
 		httpRule := x.(*apb.HttpRule)
-		rules = append(rules, httpRule)
+		if parsedRule := parseRule(httpRule); parsedRule != nil {
+			rules = append(rules, parsedRule)
 
-		// Add any additional bindings and flatten them into `rules`.
-		rules = append(rules, httpRule.GetAdditionalBindings()...)
+			// Add any additional bindings and flatten them into `rules`.
+			for _, binding := range httpRule.GetAdditionalBindings() {
+				rules = append(rules, parseRule(binding))
+			}
+		}
 	}
 
 	// Done; return the rules.
 	return rules
 }
+
+func parseRule(rule *apb.HttpRule) *HTTPRule {
+	answer := &HTTPRule{}
+	if uri := rule.GetGet(); uri != "" {
+		answer.Method = "GET"
+		answer.URI = uri
+	} else if uri := rule.GetPost(); uri != "" {
```

Comment:
Consider to write it into a `for` loop:
```go
methods := []string{"GET", "POST", "PUT", "PATCH", "DELETE", rule.GetKind()}
uris := []string{rule.GetGet(), rule.GetPost(), rule.GetPut(), rule.GetPatch(), rule.GetDelete(), rule.GetPath()}
for i := range methods {
  if uris[i] != "" {
    return &HTTPRule{
      Method: methods[i],
      URI: uris[i],
      Body: rule.GetBody(),
      ResponseBody: rule.GetResponseBody(),
    }
  }
}
return &HTTPRule{
    Body: rule.GetBody(),
    ResponseBody: rule.GetResponseBody(),
}
```

Reply:
I would rather not, for a couple of reasons:

1. It does not handle the `GetCustom()` case well (if you look at your code carefully, you will notice that it is wrong there).
2. The test case to hit the bottom return line would be weird.
3. I do not really like the "iterate over two slices that happen to be the same size" thing, and Golang has no equivalent of Python's [`itertools.zip`](https://docs.python.org/3/library/functions.html#zip).

Reply:
One alternative that is slightly better (IMO) is:

```go
	oneof := map[string]string{
		"GET":    rule.GetGet(),
		"POST":   rule.GetPost(),
		"PUT":    rule.GetPut(),
		"PATCH":  rule.GetPatch(),
		"DELETE": rule.GetDelete(),
	}
	if custom := rule.GetCustom(); custom != nil {
		oneof[custom.GetKind()] = custom.GetPath()
	}
	for method, uri := range oneof {
		if uri != "" {
			return &HTTPRule{
				Method:       method,
				URI:          uri,
				Body:         rule.GetBody(),
				ResponseBody: rule.GetResponseBody(),
			}
		}
	}
	return nil
```

My gripe with that variant is that I literally can not hit the `return nil` clause without calling the unexported method directly.

Reply:
Fair enough! Thanks!

Reply:
After thinking about it some more, I did the final variant proposed.